### PR TITLE
Added Connect & Claim functionality

### PR DESCRIPTION
**Pull request by L03TJ3**
# Description

This adds the Claim functionality.

Left on the todo's:
- Test mobile functionalities, and make responsive (Priority!)
- Handle user connecting multiple accounts through metamask
- Add proper handlers for Wallet-Connect: disconnect, connect, chainChanged, accountsChanged
- Some fine-tuning in error handling

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [ ] PR title matches follow: (Feature|Bug|Chore) Task Name
- [ ] My code follows the style guidelines of this project
- [ ] I have followed all the instructions described in the initial task (check Definitions of Done)
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,16 +8,38 @@
     "start": "next start -p $PORT"
   },
   "dependencies": {
-    "@emotion/react": "11.4.1",
-    "@emotion/styled": "11.3.0",
-    "@mui/icons-material": "5.0.0-rc.0",
-    "@mui/material": "5.0.0-rc.0",
+    "@emotion/react": "11.7.0",
+    "@emotion/styled": "11.6.0",
+    "@gooddollar/goodprotocol": "1.0.0",
```
Comment (sirpy) [inline package.json]: needs to be updated to latest

File: package.json [hunk 2]
```diff
@@ -8,16 +8,38 @@
...
+    "https-browserify": "^1.0.0",
     "merkle-tree-solidity": "1.0.8",
-    "next": "latest",
+    "next": "12.0.4",
+    "os-browserify": "^0.3.0",
     "react": "17.0.2",
-    "react-dom": "17.0.2"
+    "react-dom": "17.0.2",
+    "stream-browserify": "^3.0.0",
+    "stream-http": "^3.2.0",
```
Comment (sirpy) [inline package.json]: stream/crypto/browserify are they all needed?

File: pages/claim/claim.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+  const skipAndClaim = () => {
+    getRec(connectionDetails);
+    setQuery({status: 'claim-init'});
+  }
+  
+  const claimRep = async() => {
+    setQuery({status: 'claim-start'});
+    const claim = claimReputation(proof, connectionDetails, contractInstance);
+    claim.then((res) => {
```
Comment (sirpy) [inline pages/claim/claim.js]: dont use  then
use await

File: pages/claim/claim.js [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  const [newRecValue, setNewRecValue] = useState(null);
+
+  useEffect(() => {
+    setIsMobile(props.isMobile);
+    setProof(props.proofData);
+    getRec(props.currentConnection);
+    setConnectionDetails(props.currentConnection);
+  }, [props]);
+
+  const changeRecipient = async(e) => {
```
Comment (sirpy) [inline pages/claim/claim.js]: it is generally a good idea to wrap all functions you define in a "useCallback"
specifically functions that are passed to child components
i've wrote this comment on this function but it is relevant for all

File: pages/claim/claim.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+          variant="contained"
+          sx={{
+            mr: 1,
+            mb: isMob ? 1 : 0,
+            backgroundColor: "#9c27b0",
+           '&:hover': {
+              backgroundColor: "#60156c"
+            }
+          }}
+          onClick={() => backToSwitch()}
```
Comment (sirpy) [inline pages/claim/claim.js]: just do onClick={backToSwitch}

File: pages/claim/claim.js [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+          }}>
+            <CircularProgress color="secondary" sx={{marginTop:"20px"}} /> <br />
+            <Typography variant="span" sx={{fontStyle: 'italic'}} color="red">
+                You have a current pending transaction, please wait till confirmation.
+            </Typography>
+          </div> 
+        :
+        query.status === 'claim-start' || 
+        query.status === 'claim-success' || 
+        query.status === 'claim-failed' ?
```
Comment (sirpy) [inline pages/claim/claim.js]: can this be simplified to query.status!=='claim-init'?

File: lib/connect.serv.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+
+/**
+ * @notice Checks if a user already claimed their tokens on a network
+ * @param currentConnection contains all the necessary information of the connected wallet
+ * @return boolean's for claim-status per network
+ */
+export async function getClaimStatus(currentConnection) {
+  let contractAddressess = {
+        fuse: '0x3A9299BE789ac3730e4E4c49d6d2Ad1b8BC34DFf',
+        rootState: '0x3A9299BE789ac3730e4E4c49d6d2Ad1b8BC34DFf'
```
Comment (sirpy) [inline lib/connect.serv.js]: is this the contract address? why is it hardcoded? why is it called "rootState"?
the addresses should be read from @gooddollar/goodprotocol/releases/deploymnet.json
the keys in that json are:
staging - fuse staging contracts
staging-mainnet staging contracts on ropsten that simulate mainnet
production - fuse production contracts
production-mainnet ethereum production contracts

File: lib/connect.serv.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+ * @notice gets the current set recipient for _user from contract.
+ * @param contractInstance contract instance for connected network. is set when a user is clicking through tabs,
+ * is null on first mount.
+ * @param currentConnection necessary information of the connected wallet
+ * @return contractInstance and current recipient. Is empty address when none is set (default: _user)
+ */
+export async function getRecipient(contractInstance, currentConnection) {
+  const web3 = new Web3(new Web3.providers.HttpProvider(
+      currentConnection.chainId == 122 ? infuraConfig.pocketUrl : infuraConfig.infuraUrl)),
+      contractAddr = '0x3A9299BE789ac3730e4E4c49d6d2Ad1b8BC34DFf';
```
Comment (sirpy) [inline lib/connect.serv.js]: shouldnt be hardcoded
anything else is hardcoded?

**Review by sirpy — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (L03TJ3) [reply package.json]: no, not at all actually. can't remember why I added them

File: pages/claim/claim.js (hunk 6 above)
Comment (L03TJ3) [reply pages/claim/claim.js]: yes it can!

File: lib/connect.serv.js (hunk 7 above)
Comment (L03TJ3) [reply lib/connect.serv.js]: confused contract names and name of hash.
changed now for clarity

File: package.json (hunk 1 above)
Comment (L03TJ3) [reply package.json]: done i think:
goodprotocol "^1.0.0" ?
or should i specify the number? (1.0.2?)

File: pages/claim/claim.js (hunk 5 above)
Comment (L03TJ3) [reply pages/claim/claim.js]: done

File: lib/connect.serv.js (hunk 8 above)
Comment (L03TJ3) [reply lib/connect.serv.js]: done

File: pages/claim/claim.js [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+          getRec(connectionDetails);
+          setQuery({status: 'claim-init'});
+        }
+      }).catch((err) => {
+        setNewRecValue('0x00');
+        setQuery({status: 'init'});
+        return;
+      });
+    }
+  });
```
Comment (sirpy) [inline pages/claim/claim.js]:
```suggestion
  },[getRec, setNewRecValue, setQuery, setNewRecipient, connectionDetails]);
```

File: pages/claim/claim.js [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        const txStatus = getPendingTXStatus(currentConnection, pendingTXClaim ?? pendingTXNewRec);
+        txStatus.then((res) => {
+          if (res) {
+            setQuery({status: "claim-init"});
+            clearInterval(pendingTXStatus);
+          }
+        });
+      }, 7000);
+    }
+  }, [connectionDetails]);
```
Comment (sirpy) [inline pages/claim/claim.js]:
```suggestion
  }, [connectionDetails, setRepRecipient, setQuery]);
```

File: pages/claim/claim.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+  const backToSwitch = useCallback(() => {
+    props.toSwitch();
+  });
```
Comment (sirpy) [inline pages/claim/claim.js]:
```suggestion
  },[props.toSwitch]);
```

File: pages/claim/claim.js [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+  const backToRecipient = useCallback(() => {
+    setQuery({status: 'init'});
+  });
```
Comment (sirpy) [inline pages/claim/claim.js]:
```suggestion
  },[setQuery]);
```

File: pages/claim/claim.js [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+  const skipAndClaim = useCallback(() => {
+    getRec(connectionDetails);
+    setQuery({status: 'claim-init'});
+  }, [connectionDetails]);
```
Comment (sirpy) [inline pages/claim/claim.js]:
```suggestion
  }, [connectionDetails, setQuery, getRec]);
```

File: pages/claim/claim.js [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+    if (claim?.code) {
+      setQuery({status: 'claim-failed'});
+      setTimeout(() => {
+        setQuery({status: 'claim-init'});
+      }, 2000);
+    } else {
+      setQuery({status: 'claim-success'});
+      localStorage.removeItem("pendingClaim");
+    }
+  }, [connectionDetails, contractInstance]);
```
Comment (sirpy) [inline pages/claim/claim.js]:
```suggestion
  }, [connectionDetails, contractInstance, setQuery]);
```

File: pages/claim/claimDialog.js [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+      setConnectedAddress(null);
+      setProviderName('init');
+    } else {
+      setCurrentConnection(res);
+      setProviderEvents({status: 'init'}); 
+      setConnectedAddress(res.connectedAddress);
+      setProviderName(res.providerName);
+      setQuery({status: 'connected'});
+    }
+  });
```
Comment (sirpy) [inline pages/claim/claimDialog.js]:
```suggestion
  },[setQuery, setProviderName, setConnectedAddress, setProviderEvents, setCurrentConnection]);
```

File: pages/claim/ineligible.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+  useEffect(() => {
+    setTimeout(() => {
+      onClose();
+    }, 2250);
+  }, [onInit]);
+
+  const handleErrorClose = useCallback(() => {
+    onClose();
+  }, [props]);
```
Comment (sirpy) [inline pages/claim/ineligible.js]:
```suggestion
  }, [onClose]);
```

File: pages/claim/provider.js [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    conAddr.then((res) => {
+      setProviderInstance(res.providerInstance);
+      success(res);
+    }).catch((err) => {
+      console.log('conAddr error -->', err);
+      errorInit(err);
+    });
+  });
```
Comment (sirpy) [inline pages/claim/provider.js]:
```suggestion
  },[setQuery, setProviderInstance]);
```

File: pages/claim/switch.js [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+      }).catch((err) => {
+        // console.log('err switch -->', err);
+        if (err.code == 4902){
+          addFuseNetwork(chainId);
+        } else {
+          setError({status: null, code: err.code});
+        }
+      });
+    }  
+  }, [chainId]);
```
Comment (sirpy) [inline pages/claim/switch.js]:
```suggestion
  }, [chainId, providerInstance, isClaimed, addFuseNetwork]);
```

File: pages/claim/switch.js [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+    setError({status: 'wrongNetwork', code: 310});
+    setQuery({status: 'error'});
+    setConnectedChain('unsupported');
+    setChainId('0x00');
+  }
+
+  // Callback from claimDialog to load claim component
+  const getReputation = useCallback((chainId) => {
+    props.getRep(chainId);
+  }, [chainId]);
```
Comment (sirpy) [inline pages/claim/switch.js]:
```suggestion
  }, [chainId, props.getRep]);
```

File: pages/index.js [hunk 20]
```diff
@@ -131,14 +135,28 @@ const AirdropData = ({ hexProof, proofIndex, addr, reputationInWei }) => {
...
-    console.log({ result });
+    const result = await fetch(`/api/repAirdrop/${addr}`)
+                   .then((_) => _.json());
     setData(result);
+    result.error ? setErrorOpen(true) : setOpen(true);
   };
 
+  const handleClose = useCallback((value) => {
+    setOpen(false);
+  });
```
Comment (sirpy) [inline pages/index.js]:
```suggestion
  }, [setOpen]);
```

File: pages/index.js [hunk 21]
```diff
@@ -131,14 +135,28 @@ const AirdropData = ({ hexProof, proofIndex, addr, reputationInWei }) => {
...
+
+  const handleErrorClose = useCallback((value) => {
+    setErrorOpen(false);
+  })
```
Comment (sirpy) [inline pages/index.js]:
```suggestion
  },[setErrorOpen])
```

**Review by sirpy — CHANGES_REQUESTED** [review]
- always use square brackets, even if empty, otherwise function will be redefined everytime.
empty brackets means "on component mount"
- in the brackets you need to include all component variables that the function uses, because on re-render any of them might change. so if one of them changes function needs to be redefined
- you still have some methods you didnt wrap in usecallback

**Pull request merged**